### ISSUE TITLE
EvenJournalReader.readFromEventJournal() doesn't complete with concurrent readers

### ISSUE THREAD

**Issue by cangencer** [hazelcast/hazelcast#13498]
On version 3.10.3, it seems that sometimes the read operation doesn't return in a timely when there are multiple concurrent readers.

Test below passes when the marked line is commented out. 

```
package test;

import com.hazelcast.config.Config;
import com.hazelcast.config.EventJournalConfig;
import com.hazelcast.core.EntryEventType;
import com.hazelcast.core.ExecutionCallback;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import com.hazelcast.internal.journal.EventJournalReader;
import com.hazelcast.map.journal.EventJournalMapEvent;
import com.hazelcast.projection.Projections;
import com.hazelcast.ringbuffer.ReadResultSet;
import com.hazelcast.spi.properties.GroupProperty;
import com.hazelcast.test.AssertTask;
import com.hazelcast.test.HazelcastParallelClassRunner;
import com.hazelcast.test.HazelcastTestSupport;
import com.hazelcast.test.annotation.Repeat;
import com.hazelcast.util.function.Consumer;
import com.hazelcast.util.function.Predicate;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Arrays;
import java.util.HashSet;
import java.util.concurrent.CopyOnWriteArrayList;

import static org.junit.Assert.assertEquals;

@RunWith(HazelcastParallelClassRunner.class)
public class EventJournalTest extends HazelcastTestSupport {
    private static final int JOURNAL_CAPACITY = 1000;
    private IMap<String, Integer> map;
    private String MAP_NAME = "test";
    private EventJournalReader<EventJournalMapEvent<String, Integer>> reader;
    private HazelcastInstance hz;

    @Before
    public void setUp() {
        Config config = new Config();

        EventJournalConfig journalConfig = new EventJournalConfig()
                .setMapName("*")
                .setCapacity(JOURNAL_CAPACITY)
                .setEnabled(true);
        config.addEventJournalConfig(journalConfig);
        config.setProperty("hazelcast.partition.count", "1");
        config.setProperty(GroupProperty.OPERATION_CALL_TIMEOUT_MILLIS.getName(), "60000");
        hz = this.createHazelcastInstance(config);

        map = hz.getMap(MAP_NAME);
        reader = (EventJournalReader<EventJournalMapEvent<String, Integer>>) map;
    }

    @Test
    @Repeat(1000)
    public void test() {
        for (int i = 1; i <= 3; i++) {
            map.put("1", i);
        }

        // test passes when this line is commented out
        readFromJournal(0, e -> e.getType() == EntryEventType.ADDED, e -> { });

        CopyOnWriteArrayList<Integer> updates = new CopyOnWriteArrayList<>();
        readFromJournal(0, e -> e.getType() == EntryEventType.UPDATED, e -> {
            updates.add(e.getNewValue());
            if (e.getNewValue() < 100) {
                map.put(e.getKey(), e.getNewValue() * 100);
            }
        });

        assertTrueEventually(new AssertTask() {
            @Override
            public void run() {
                System.out.println(updates);
                assertEquals(new HashSet<>(Arrays.asList(2, 3, 200, 300)), new HashSet<>(updates));
            }
        });
    }

    private void readFromJournal(long seq,
                                 Predicate<EventJournalMapEvent<String, Integer>> predicate,
                                 Consumer<EventJournalMapEvent<String, Integer>> consumer
    ) {
        reader.readFromEventJournal(
                seq,
                1,
                16,
                0,
                predicate, Projections.identity()
        ).andThen(new ExecutionCallback<ReadResultSet<EventJournalMapEvent<String, Integer>>>() {
            @Override
            public void onResponse(ReadResultSet<EventJournalMapEvent<String, Integer>> response) {
                readFromJournal(response.getNextSequenceToReadFrom(), predicate, consumer);
                for (EventJournalMapEvent<String, Integer> event : response) {
                    consumer.accept(event);
                }
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        });
    }
}
```

**Comment by lukasherman** [comment]
Does it mean that bypassing the predicate might be a workaround?

**Comment by cangencer** [comment]
@lukasherman Yes, at least in the example above, the problem doesn't happen if you change the Predicate from `ADDED` to `UPDATED`.

**Comment by mmedenjak** [comment]
This is happening basically because of the same reason why this issue was happening - https://github.com/hazelcast/hazelcast/pull/11310.
Waiting operations (e.g. read operations) are kept in queues and must be processed in order. Initially, the queue was common for all partitions for a single event journal (the previous issue). That meant that an operation reading from one partition could block an another operation reading from a different partition. This was fixed by having the partition ID part of the "wait key". Now operations reading from different partitions can read concurrently.
The problem is that now we have two operations on the same partition, with a different predicate. The operation waiting for ADDED events is blocking the operation waiting on UPDATED events.
We might add the predicate as part of the "wait key" but that's too much. Most wait queues would be consisted of just a single operation and we wouldn't be able to efficiently determine which wait queues to "wake up" without calling all of the predicates.

The problem here is basically that the ringbuffer read operation is abusing the wait-notify mechanism to allow predicates and `minSize` read items. We should stop using the wait-notify mechanism and have the ringbuffer service be responsible for keeping all of the pending read operations and query each of them on every event (like in a regular pub-sub system). Unfortunately this is a bigger change and it won't make it into 3.11.

**Comment by lukasherman** [comment]
So far `streamMapP` in Jet works correctly with the same `predicateFn` with filtering applied in `projectionFn` by simply returning 'null`, even under moderate load.

**Comment by lukasherman** [comment]
Observed 3-12 seconds occasional (cca 0.4% cases) delays between object added/change observed by event journal, under intense load (cca 70%-90% system capacity). The flow eventually completes, but overall system SLA is breached.

**Comment by mmedenjak** [comment]
Hi @lukasherman !

The operation which adds an entry to the map will also synchronously add an entry to the event journal and will resume any pending journal read operation (also in the same thread and in the same method call). So in most cases you should receive the response for the event journal read operation as soon as the map is mutated.

Is it possible that we are encountering the same issue where the first pending journal read operation has a non-matching predicate and a journal read operation with a matching predicate is deeper in the queue? Are you using different predicates when reading from the journal?

Is it possible the application is suspended for some reason (e.g. GC pause)? Can you profile your app?

**Comment by lukasherman** [comment]
Hi @mmedenjak ,
profiling shows that GC pauses are within ms range, even full GC < 150ms. We have already refactored the workflows with single journal per map approach, moved non-critical and dependent code to local map listener instead, and so far it works correctly. Instead of using multiple journal readers per map we redesigned the flow to use single reader and filter / split the data based on content, building DAG dynamically during initialization.

**Comment by mmedenjak** [comment]
How do you measure the latency between object add and change observed? Can you post your code here? Can you share a small reproducer?

**Comment by lukasherman** [comment]
The measurement is based on internal properties of the object before `put` - field `created` is pre-initialized to `Instant.now()`.
This behavior was observed during performance testing of a quite complex application, so extracting a reproducer might be complicated. Will think about that.
Is it possible that a non-matching event journal predicate (UPDATE) on a second map might affect matching predicate (ADDED) on a first map not being resolved in time, is the reader somehow shared? Because that was basically our case.

**Comment by lukasherman** [comment]
This is still one of the major Jet pain points. I find myself designing workarounds for this issue all the time.

**Comment by wfhartford** [comment]
I just ran into this bug in version 5.2.3. I understand that the _correct_ fix is quite a bit of engineering in hazelcast internals which I don't understand, but would it be possible to implement a workaround in hazelcast? It seems like the recommended workaround for users to apply is to more the predicate logic into the projection function. This is the direction I will be going in my application. Would it be possible to apply this workaround in Hazelcast itself? The user supplied predicate and the projection function could be combined into a new projection function under the covers, the new projection function could then be used the the always true predicate. When the proper fix for this bug is ready, that workaround could be removed and users would benefit without having to modify their code.

In my quest to understand the bug I was experiencing before I found this issue, I built a small tool to help me understand the issue. The tool is available at https://github.com/wfhartford/hazelcast-journal-stream in case it might be of use to anyone.

**Comment by lukasherman** [comment]
fixed in v5.7.0, commit dd61fb8d263c4c5b645224dab273f4f61601705d
